ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs

```diff
@@ -265,34 +272,114 @@ newResolverResource resolvConf = go
      | TraceLocalRootWaiting DomainAddress DiffTime
      | TraceLocalRootResult  DomainAddress [(IPv4, DNS.TTL)]
      | TraceLocalRootFailure DomainAddress DNSorIOError
        --TODO: classify DNS errors, config error vs transitory
   deriving Show
 
-
 -- |
 --
-localRootPeersProvider :: Tracer IO TraceLocalRootPeers
-                       -> TimeoutFn IO
-                       -> DNS.ResolvConf
-                       -> StrictTVar IO (Map DomainAddress (Map Socket.SockAddr PeerAdvertise))
-                       -> NonEmpty (DomainAddress, PeerAdvertise)
-                       -> IO Void
-localRootPeersProvider tracer timeout resolvConf rootPeersVar domains = do
-    traceWith tracer (TraceLocalRootDomains (NonEmpty.toList domains))
+localRootPeersProvider
+  :: Tracer IO TraceLocalRootPeers
+  -> TimeoutFn IO
+  -> DNS.ResolvConf
+  -> StrictTVar IO [(Int, Map Socket.SockAddr PeerAdvertise)]
+  -> StrictTVar IO [(Int, Map RelayAddress PeerAdvertise)]
+  -> IO Void
+localRootPeersProvider tracer
+                       timeout
+                       resolvConf
+                       rootPeersGroupsVar
+                       domainsGroupsVar = do
+  domainsGroups <- atomically $ readTVar domainsGroupsVar
+  traceWith tracer (TraceLocalRootDomains domainsGroups)
 #if !defined(mingw32_HOST_OS)
-    rr <- asyncResolverResource resolvConf
+  rr <- asyncResolverResource resolvConf
 #else
-    let rr = newResolverResource resolvConf
+  let rr = newResolverResource resolvConf
 #endif
-    withAsyncAll (map (monitorDomain rr) (NonEmpty.toList domains)) $ \asyncs ->
-      snd <$> waitAny asyncs
+  let
+      -- Flatten the local root peers groups and associate a group index to each
+      -- DomainAddress to be monitorized.
+      addresses = [ (group, address, pa)
+                  | (group, (_, m)) <- zip [0..] domainsGroups
+                  , (address, pa) <- Map.toList m ]
+      domains = [ (group, domain, pa)
+                | (group, RelayDomain domain, pa) <- addresses ]
+      ipsAddr = [ (group, IP.toSockAddr (ip, port), pa)
+                | (group, RelayAddress ip port, pa) <- addresses ]
+
+      -- Since we want to preserve the number of groups, the targets, and
+      -- the addresses within each group, we fill the TVar with
+      -- a placeholder list, in order for each monitored DomainAddress to
+      -- be updated in the correct group.
+      placeholder = map (\(t, _) -> (t, Map.empty)) domainsGroups
+      placeholder' =
+        foldr (\(group, addr, pa) ph
+                -> let (target, entry) = ph !! group
+                   in updateList group (target, Map.insert addr pa entry) ph)
+              placeholder
+              ipsAddr
+
+  atomically $
+    writeTVar rootPeersGroupsVar placeholder'
+
+  -- Launch DomainAddress monitoring threads and
+  -- wait for threads to return or for local config changes
+  withAsyncAll (map (monitorDomain rr) domains) $ \as ->
+    atomically $ do
+      if null as
```

Comment:
Is this necessary? what happens if I wait on an empty list?

Reply:
looking at the implementation of `withAsyncAll` it will run the action with `as = []`; `waitAnySTM` will block on the empty list: https://hackage.haskell.org/package/async-2.2.3/docs/src/Control.Concurrent.Async.html#waitAny

Reply:
Does that mean I can simplify this code?